File: app/src/main/java/org/puder/trs80/keyboard/JoystickView.java
/*
 * Copyright 2012-2013, Arno Puder
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.puder.trs80.keyboard;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.Path;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import org.puder.trs80.EmulatorActivity;

/**
 * http://www.trs-80.com/wordpress/zaps-patches-pokes-tips/internals/#keyboard13
 */
public class JoystickView extends View {

    private boolean           leftKeyPressed    = false;
    private boolean           rightKeyPressed   = false;
    private boolean           upKeyPressed      = false;
    private boolean           downKeyPressed    = false;

    private float             joystickX         = -1;
    private float             joystickY         = -1;

    private float             radiusButton;
    private float             radiusJoystick;
    private float             circleStrokeWidth = 10;
    private float             triangleSideLength;
    private Bitmap            joystickBitmap;
    private boolean           joystickIsPressed;

    protected Paint           paint;
    protected KeyboardManager keyboardManager;

    public JoystickView(Context context, AttributeSet attrs) {
        super(context, attrs);

        EmulatorActivity emulator = (EmulatorActivity) context;
        this.keyboardManager = emulator.getKeyboardManager();

        paint = new Paint();
        paint.setAntiAlias(true);

        radiusButton = pxFromDp(40);
        triangleSideLength = pxFromDp(15);

        this.setOnTouchListener(new OnTouchListener() {

            @Override
            public boolean onTouch(View view, MotionEvent event) {
                int action = event.getAction() & MotionEvent.ACTION_MASK;
                if (action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_CANCEL) {
                    unpressAllKeys();
                    joystickX = -1;
                    joystickY = -1;
                    joystickIsPressed = false;
                    invalidate();
                    return true;
                }

                joystickIsPressed = true;
                joystickX = event.getX();
                joystickY = event.getY();
                invalidate();

                float cx = getWidth() / 2.0f;
                float cy = getHeight() / 2.0f;
                double dx = joystickX - cx;
                double dy = cy - joystickY;
                pressKeys(dx, dy);
                return true;
            }
        });
    }

    private void pressKeyDown() {
        keyboardManager.pressKeyDown();
    }

    private void pressKeyUp() {
        keyboardManager.pressKeyUp();
    }

    private void pressKeyLeft() {
        keyboardManager.pressKeyLeft();
    }

    private void pressKeyRight() {
        keyboardManager.pressKeyRight();
    }

    private void unpressKeyDown() {
        keyboardManager.unpressKeyDown();
    }

    private void unpressKeyUp() {
        keyboardManager.unpressKeyUp();
    }

    private void unpressKeyLeft() {
        keyboardManager.unpressKeyLeft();
    }

    private void unpressKeyRight() {
        keyboardManager.unpressKeyRight();
    }

    private float pxFromDp(float dp) {
        return dp * this.getContext().getResources().getDisplayMetrics().density;
    }

    private void unpressAllKeys() {
        if (leftKeyPressed) {
            leftKeyPressed = false;
            unpressKeyLeft();
        }
        if (rightKeyPressed) {
            rightKeyPressed = false;
            unpressKeyRight();
        }
        if (upKeyPressed) {
            upKeyPressed = false;
            unpressKeyUp();
        }
        if (downKeyPressed) {
            downKeyPressed = false;
            unpressKeyDown();
        }
    }

    private void pressKeys(double dx, double dy) {
        if (Math.sqrt(dx * dx + dy * dy) < radiusButton) {
            unpressAllKeys();
            return;
        }

        double angle = computeAngle(dx, dy);
        final double slice = 360 / 8;
        final double start = slice / 2;

        // Right key
        if (angle < start || angle >= start + 7 * slice) {
            if (!rightKeyPressed) {
                rightKeyPressed = true;
                pressKeyRight();
            }
            if (leftKeyPressed) {
                leftKeyPressed = false;
                unpressKeyLeft();
            }
            if (upKeyPressed) {
                upKeyPressed = false;
                unpressKeyUp();
            }
            if (downKeyPressed) {
                downKeyPressed = false;
                unpressKeyDown();
            }
            return;
        }

        // Right & up keys
        if (angle < start + slice) {
            if (!rightKeyPressed) {
                rightKeyPressed = true;
                pressKeyRight();
            }
            if (leftKeyPressed) {
                leftKeyPressed = false;
                unpressKeyLeft();
            }
            if (!upKeyPressed) {
                upKeyPressed = true;
                pressKeyUp();
            }
            if (downKeyPressed) {
                downKeyPressed = false;
                unpressKeyDown();
            }
            return;
        }

        // Up key
        if (angle < start + 2 * slice) {
            if (rightKeyPressed) {
                rightKeyPressed = false;
                unpressKeyRight();
            }
            if (leftKeyPressed) {
                leftKeyPressed = false;
                unpressKeyLeft();
            }
            if (!upKeyPressed) {
                upKeyPressed = true;
                pressKeyUp();
            }
            if (downKeyPressed) {
                downKeyPressed = false;
                unpressKeyDown();
            }
            return;
        }

        // Up & left keys
        if (angle < start + 3 * slice) {
            if (rightKeyPressed) {
                rightKeyPressed = false;
                unpressKeyRight();
            }
            if (!leftKeyPressed) {
                leftKeyPressed = true;
                pressKeyLeft();
            }
            if (!upKeyPressed) {
                upKeyPressed = true;
                pressKeyUp();
            }
            if (downKeyPressed) {
                downKeyPressed = false;
                unpressKeyDown();
            }
            return;
        }

        // Left key
        if (angle < start + 4 * slice) {
            if (rightKeyPressed) {
                rightKeyPressed = false;
                unpressKeyRight();
            }
            if (!leftKeyPressed) {
                leftKeyPressed = true;
                pressKeyLeft();
            }
            if (upKeyPressed) {
                upKeyPressed = false;
                unpressKeyUp();
            }
            if (downKeyPressed) {
                downKeyPressed = false;
                unpressKeyDown();
            }
            return;
        }

        // Left & down keys
        if (angle < start + 5 * slice) {
            if (rightKeyPressed) {
                rightKeyPressed = false;
                unpressKeyRight();
            }
            if (!leftKeyPressed) {
                leftKeyPressed = true;
                pressKeyLeft();
            }
            if (upKeyPressed) {
                upKeyPressed = false;
                unpressKeyUp();
            }
            if (!downKeyPressed) {
                downKeyPressed = true;
                pressKeyDown();
            }
            return;
        }

        // Down key
        if (angle < start + 6 * slice) {
            if (rightKeyPressed) {
                rightKeyPressed = false;
                unpressKeyRight();
            }
            if (leftKeyPressed) {
                leftKeyPressed = false;
                unpressKeyLeft();
            }
            if (upKeyPressed) {
                upKeyPressed = false;
                unpressKeyUp();
            }
            if (!downKeyPressed) {
                downKeyPressed = true;
                pressKeyDown();
            }
            return;
        }

        // Down & right keys
        if (angle < start + 7 * slice) {
            if (!rightKeyPressed) {
                rightKeyPressed = true;
                pressKeyRight();
            }
            if (leftKeyPressed) {
                leftKeyPressed = false;
                unpressKeyLeft();
            }
            if (upKeyPressed) {
                upKeyPressed = false;
                unpressKeyUp();
            }
            if (!downKeyPressed) {
                downKeyPressed = true;
                pressKeyDown();
            }
        }
    }

    private double computeAngle(double dx, double dy) {
        if (dx == 0) {
            return dy >= 0 ? 90 : 270;
        }
        double atan = Math.toDegrees(Math.atan(dy / dx));
        if (dx < 0 && dy > 0) {
            return 180 + atan;
        }
        if (dx < 0 && dy < 0) {
            return 180 + atan;
        }
        if (dx > 0 && dy < 0) {
            return 360 + atan;
        }
        return atan;
    }

    @Override
    public void onDraw(Canvas canvas) {
        int height = this.getHeight();
        int width = this.getWidth();
        float cx = width / 2.0f;
        float cy = height / 2.0f;
        if (joystickX == -1) {
            joystickX = cx;
        }
        if (joystickY == -1) {
            joystickY = cy;
        }
        createJoystickBitmap();

        paint.setColor(joystickIsPressed ? Color.LTGRAY : Color.GRAY);
        paint.setAlpha(100);
        canvas.drawBitmap(joystickBitmap, 0, 0, paint);

        float x = joystickX - cx;
        float y = joystickY - cy;
        double dist = Math.sqrt(x * x + y * y);
        if (dist > radiusJoystick) {
            canvas.translate(cx, cy);
            double dx = joystickX - cx;
            double dy = cy - joystickY;
            canvas.rotate(-(float) computeAngle(dx, dy));
            canvas.drawCircle(radiusJoystick, 0, radiusButton, paint);
        } else {
            canvas.drawCircle(joystickX, joystickY, radiusButton, paint);
        }
    }

    private void createJoystickBitmap() {
        int height = this.getHeight();
        int width = this.getWidth();
        if (joystickBitmap != null && joystickBitmap.getWidth() == width
                && joystickBitmap.getHeight() == height) {
            return;
        }
        joystickBitmap = Bitmap.createBitmap(width, height, Config.ARGB_8888);
        Canvas canvas = new Canvas(joystickBitmap);
        radiusJoystick = (width - circleStrokeWidth) / 2 - radiusButton;
        float cx = width / 2.0f;
        float cy = height / 2.0f;
        Paint paint = new Paint();
        paint.setAntiAlias(true);
        paint.setColor(Color.GRAY);
        paint.setStyle(Style.STROKE);
        paint.setStrokeWidth(circleStrokeWidth);
        canvas.drawCircle(cx, cy, radiusJoystick, paint);
        paint.setStrokeWidth(1);
        canvas.drawCircle(cx, cy, radiusButton, paint);

        paint.setStyle(Style.FILL_AND_STROKE);
        canvas.save();
        canvas.translate(cx, cy);
        Path path = new Path();
        path.reset();
        path.moveTo(radiusJoystick, -triangleSideLength / 2.0f);
        path.lineTo(radiusJoystick, triangleSideLength / 2.0f);
        path.lineTo(radiusJoystick + triangleSideLength * 0.667f, 0);
        path.close();
        for (int i = 0; i < 8; i++) {
            canvas.drawPath(path, paint);
            canvas.rotate(45);
        }
        canvas.restore();
    }
}

File: app/src/main/java/org/puder/trs80/keyboard/JoystickView.kt
/*
 * Copyright The TRS-80 App Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.puder.trs80.keyboard

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import org.puder.trs80.EmulatorActivity
import kotlin.math.atan
import kotlin.math.sqrt

/**
 * An eight-way on-screen joystick that drives the TRS-80 cursor keys.
 *
 * The circle around the centre is a dead zone; outside of it the angle towards the touch point
 * is quantised into eight 45 degree slices, the four diagonal ones pressing two cursor keys at
 * once. The hosting context must be an [EmulatorActivity].
 *
 * See http://www.trs-80.com/wordpress/zaps-patches-pokes-tips/internals/#keyboard13
 */
@SuppressLint("ClickableViewAccessibility")
class JoystickView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyle: Int = 0) : View(context, attrs, defStyle) {

    private val keyboardManager = (context as EmulatorActivity).keyboardManager
    private val paint = Paint().apply { isAntiAlias = true }

    private val radiusButton = pxFromDp(40f)
    private val triangleSideLength = pxFromDp(15f)

    private var leftKeyPressed = false
    private var rightKeyPressed = false
    private var upKeyPressed = false
    private var downKeyPressed = false

    private var joystickX = -1f
    private var joystickY = -1f
    private var joystickIsPressed = false

    private var radiusJoystick = 0f
    private var joystickBitmap: Bitmap? = null

    init {
        setOnTouchListener { _, event ->
            handleTouch(event)
            true
        }
    }

    override fun onDraw(canvas: Canvas) {
        val cx = width / 2.0f
        val cy = height / 2.0f
        if (joystickX == -1f) {
            joystickX = cx
        }
        if (joystickY == -1f) {
            joystickY = cy
        }
        val bitmap = obtainJoystickBitmap()

        paint.color = if (joystickIsPressed) Color.LTGRAY else Color.GRAY
        paint.alpha = 100
        canvas.drawBitmap(bitmap, 0f, 0f, paint)

        val x = joystickX - cx
        val y = joystickY - cy
        if (sqrt((x * x + y * y).toDouble()) > radiusJoystick) {
            // Outside the ring: pin the knob to the ring, in the direction of the touch.
            canvas.translate(cx, cy)
            val angle = computeAngle((joystickX - cx).toDouble(), (cy - joystickY).toDouble())
            canvas.rotate(-angle.toFloat())
            canvas.drawCircle(radiusJoystick, 0f, radiusButton, paint)
        } else {
            canvas.drawCircle(joystickX, joystickY, radiusButton, paint)
        }
    }

    private fun handleTouch(event: MotionEvent) {
        val action = event.action and MotionEvent.ACTION_MASK
        if (action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_CANCEL) {
            unpressAllKeys()
            joystickX = -1f
            joystickY = -1f
            joystickIsPressed = false
            invalidate()
            return
        }

        joystickIsPressed = true
        joystickX = event.x
        joystickY = event.y
        invalidate()

        val cx = width / 2.0f
        val cy = height / 2.0f
        pressKeys((joystickX - cx).toDouble(), (cy - joystickY).toDouble())
    }

    /**
     * Maps the offset of the touch point from the centre onto the cursor keys. [dx] grows to the
     * right and [dy] grows upwards, both in pixels.
     */
    private fun pressKeys(dx: Double, dy: Double) {
        if (sqrt(dx * dx + dy * dy) < radiusButton) {
            // Inside the dead zone around the centre.
            unpressAllKeys()
            return
        }

        val angle = computeAngle(dx, dy)
        when {
            angle < START || angle >= START + 7 * SLICE ->
                setKeys(right = true, left = false, up = false, down = false)

            angle < START + SLICE ->
                setKeys(right = true, left = false, up = true, down = false)

            angle < START + 2 * SLICE ->
                setKeys(right = false, left = false, up = true, down = false)

            angle < START + 3 * SLICE ->
                setKeys(right = false, left = true, up = true, down = false)

            angle < START + 4 * SLICE ->
                setKeys(right = false, left = true, up = false, down = false)

            angle < START + 5 * SLICE ->
                setKeys(right = false, left = true, up = false, down = true)

            angle < START + 6 * SLICE ->
                setKeys(right = false, left = false, up = false, down = true)

            else ->
                setKeys(right = true, left = false, up = false, down = true)
        }
    }

    /**
     * Brings the four cursor keys into the requested state, sending a key event only for those
     * that actually change.
     */
    private fun setKeys(right: Boolean, left: Boolean, up: Boolean, down: Boolean) {
        if (rightKeyPressed != right) {
            rightKeyPressed = right
            if (right) keyboardManager.pressKeyRight() else keyboardManager.unpressKeyRight()
        }
        if (leftKeyPressed != left) {
            leftKeyPressed = left
            if (left) keyboardManager.pressKeyLeft() else keyboardManager.unpressKeyLeft()
        }
        if (upKeyPressed != up) {
            upKeyPressed = up
            if (up) keyboardManager.pressKeyUp() else keyboardManager.unpressKeyUp()
        }
        if (downKeyPressed != down) {
            downKeyPressed = down
            if (down) keyboardManager.pressKeyDown() else keyboardManager.unpressKeyDown()
        }
    }

    private fun unpressAllKeys() {
        if (leftKeyPressed) {
            leftKeyPressed = false
            keyboardManager.unpressKeyLeft()
        }
        if (rightKeyPressed) {
            rightKeyPressed = false
            keyboardManager.unpressKeyRight()
        }
        if (upKeyPressed) {
            upKeyPressed = false
            keyboardManager.unpressKeyUp()
        }
        if (downKeyPressed) {
            downKeyPressed = false
            keyboardManager.unpressKeyDown()
        }
    }

    /** @return The angle of ([dx], [dy]) in degrees, counter-clockwise from the positive x axis. */
    private fun computeAngle(dx: Double, dy: Double): Double {
        if (dx == 0.0) {
            return if (dy >= 0) 90.0 else 270.0
        }
        val atan = Math.toDegrees(atan(dy / dx))
        return when {
            // Note that dy == 0.0 with dx < 0 deliberately falls through to `atan`, i.e. reports
            // 0 rather than 180 degrees. Preserved from the original implementation.
            dx < 0 && dy != 0.0 -> 180 + atan
            dx > 0 && dy < 0 -> 360 + atan
            else -> atan
        }
    }

    /** @return The cached backdrop bitmap, redrawing it if the view has been resized. */
    private fun obtainJoystickBitmap(): Bitmap {
        joystickBitmap?.let {
            if (it.width == width && it.height == height) {
                return it
            }
        }
        return createJoystickBitmap().also { joystickBitmap = it }
    }

    /** Draws the ring, the dead zone circle and the eight direction arrows. */
    private fun createJoystickBitmap(): Bitmap {
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        radiusJoystick = (width - CIRCLE_STROKE_WIDTH) / 2 - radiusButton
        val cx = width / 2.0f
        val cy = height / 2.0f

        val bitmapPaint = Paint().apply {
            isAntiAlias = true
            color = Color.GRAY
            style = Paint.Style.STROKE
            strokeWidth = CIRCLE_STROKE_WIDTH
        }
        canvas.drawCircle(cx, cy, radiusJoystick, bitmapPaint)
        bitmapPaint.strokeWidth = 1f
        canvas.drawCircle(cx, cy, radiusButton, bitmapPaint)

        bitmapPaint.style = Paint.Style.FILL_AND_STROKE
        canvas.save()
        canvas.translate(cx, cy)
        val arrow = Path().apply {
            moveTo(radiusJoystick, -triangleSideLength / 2.0f)
            lineTo(radiusJoystick, triangleSideLength / 2.0f)
            lineTo(radiusJoystick + triangleSideLength * 0.667f, 0f)
            close()
        }
        repeat(8) {
            canvas.drawPath(arrow, bitmapPaint)
            canvas.rotate(45f)
        }
        canvas.restore()
        return bitmap
    }

    private fun pxFromDp(dp: Float): Float = dp * context.resources.displayMetrics.density

    private companion object {
        private const val CIRCLE_STROKE_WIDTH = 10f

        /** The joystick is divided into eight equal slices, one per direction. */
        private const val SLICE = 45.0

        /** Half a slice, so that the first slice is centred on the positive x axis. */
        private const val START = SLICE / 2
    }
}
